src/main/kotlin/org/dif/crypto/JWE.kt

```diff
@@ -0,0 +1,155 @@
+package org.dif.crypto
+
+import com.nimbusds.jose.EncryptionMethod
+import com.nimbusds.jose.JOSEObjectType
+import com.nimbusds.jose.JWEAlgorithm
+import com.nimbusds.jose.JWEHeader
+import com.nimbusds.jose.JWEObjectJSON
+import com.nimbusds.jose.Payload
+import com.nimbusds.jose.UnprotectedHeader
+import com.nimbusds.jose.crypto.ECDH1PUDecrypterMulti
+import com.nimbusds.jose.crypto.ECDH1PUEncrypterMulti
+import com.nimbusds.jose.crypto.ECDH1PUX25519DecrypterMulti
+import com.nimbusds.jose.crypto.ECDH1PUX25519EncrypterMulti
+import com.nimbusds.jose.crypto.ECDHDecrypterMulti
+import com.nimbusds.jose.crypto.ECDHEncrypterMulti
+import com.nimbusds.jose.crypto.X25519DecrypterMulti
+import com.nimbusds.jose.crypto.X25519EncrypterMulti
+import com.nimbusds.jose.jwk.ECKey
+import com.nimbusds.jose.jwk.OctetKeyPair
+import com.nimbusds.jose.util.Base64URL
+import com.nimbusds.jose.util.Pair
+import org.dif.common.AnonCryptAlg
+import org.dif.common.AuthCryptAlg
+import org.dif.common.Typ
+import org.dif.crypto.key.Key
+import org.dif.exceptions.MalformedMessageException
+import org.dif.exceptions.UnsupportedAlgorithm
+import org.dif.exceptions.UnsupportedJWKException
+import org.dif.utils.asKeys
+import org.dif.utils.component1
+import org.dif.utils.component2
+import java.security.MessageDigest
+
+fun authEncrypt(payload: String, auth: AuthCryptAlg, from: Key, to: List<Key>): EncryptResult {
+    val digest = MessageDigest.getInstance("SHA-256")
+
+    val skid = from.id
+    val kids = to.map { it.id }.sorted()
+
+    val apu = Base64URL.encode(from.id)
+    val apv = Base64URL.encode(digest.digest(kids.joinToString(".").encodeToByteArray()))
+
+    val (alg, enc) = when (auth) {
+        AuthCryptAlg.A256CBC_HS512_ECDH_1PU_A256KW -> Pair(JWEAlgorithm.ECDH_1PU_A256KW, EncryptionMethod.A256CBC_HS512)
+    }
+
+    val jweHeader = JWEHeader.Builder(alg, enc)
+        .type(JOSEObjectType(Typ.Encrypted.typ))
+        .agreementPartyVInfo(apv)
+        .agreementPartyUInfo(apu)
+        .senderKeyID(skid)
+        .build()
+
+    val sender = from.jwk
+    val recipients = to.map { Pair.of(UnprotectedHeader.Builder(it.id).build(), it.jwk) }
+
+    val encryptor = when (sender) {
+        is ECKey -> ECDH1PUEncrypterMulti(sender, recipients.asKeys())
+        is OctetKeyPair -> ECDH1PUX25519EncrypterMulti(sender, recipients.asKeys())
+        else -> throw UnsupportedJWKException(sender.javaClass.name)
+    }
+
+    return JWEObjectJSON(jweHeader, Payload(Base64URL.encode(payload)))
+        .apply { encrypt(encryptor) }
```

Comment:
What if there is exception raised by Nimbus? I think we need to re-wrap it into `DIDCommException` (same for JWS).

Reply:
Yes, I suggest add exception handling next PR, because it's not clear for me. (Which exceptions should be supported)

Reply:
a `sealed class` for the known errors could be nice too - users probably don't need the whole stack with `MalformedMessageException` for an error during decryption